src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// </summary>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            var newName = newDocumentInfo.Name;
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
+                return new RenameDocumentResult(solution, renameSymbolSucceeded: false);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldRenameType)
+            {
+                solution = await RenameTypeToMatchDocumentNameAsync(document.Id, solution, newName, document.Name, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
+            var oldNameWithoutExtension = Path.GetFileNameWithoutExtension(originalName);
+
+            var document = solution.GetDocument(documentId);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
```

Comment:
we shoudl probably validate that the public entrypoint is called on a doc containing symbols that can be renamed.  i.e. not a TS or Xaml file.

Reply:
Is `document.SupportsSyntaxTree` a good check for that?

Reply:
yes.  that would work for me.